ttnn/cpp/ttnn/operations/pool/upsample/device/kernels/dataflow/reader_bilinear_multi_core_sharded.cpp

```diff
@@ -0,0 +1,108 @@
+// SPDX-FileCopyrightText: © 2023 Tenstorrent Inc.
+//
+// SPDX-License-Identifier: Apache-2.0
+
+#include "dataflow_api.h"
+#include "ttnn/cpp/ttnn/deprecated/tt_dnn/kernels/dataflow/moreh_common.hpp"
+
+#define ALWI inline __attribute__((always_inline))
+
+// Fill given four values into the memory starting at the given address.
+// WARNING: Use with caution as there's no memory protection. Make sure size is within limits
+ALWI bool fill_four_val(uint32_t begin_addr, uint16_t val, uint16_t val1, uint16_t val2, uint16_t val3) {
+    volatile tt_l1_ptr uint32_t* ptr = reinterpret_cast<volatile tt_l1_ptr uint32_t*>(begin_addr);
+
+    ptr[0] = (val | (val1 << 16));
+    ptr[1] = (val2 | (val3 << 16));
+    return true;
+}
+
+
+void kernel_main() {
+
+    uint32_t stick_nbytes = get_arg_val<uint32_t>(0);
+    uint32_t in_image_rows_per_core = get_arg_val<uint32_t>(1);
+    uint32_t scale_h = get_arg_val<uint32_t>(2);
+    uint32_t scale_w = get_arg_val<uint32_t>(3);
+    uint32_t in_w = get_arg_val<uint32_t>(4);
+    uint32_t out_w = get_arg_val<uint32_t>(5);
+    uint32_t src1_addr  = get_arg_val<uint32_t>(6);
+    uint32_t read_offset = get_arg_val<uint32_t>(8);
+    uint32_t is_last_row = get_arg_val<uint32_t>(9);
+    uint32_t in_h = 1;
+    constexpr bool src1_is_dram = false;
+
+    DPRINT << "scale_h " << scale_h << " scale_w " << scale_w << " in_w " << in_w << " out_w " << out_w << " stick_nbytes " << stick_nbytes << ENDL();
+    constexpr uint32_t in_cb_id = get_compile_time_arg_val(0);
+    constexpr uint32_t out_cb_id = tt::CB::c_in1;
+    constexpr uint32_t is_reader = get_compile_time_arg_val(2);
+
+    uint32_t in_image_row_nbytes = in_w * stick_nbytes;
+    uint32_t out_image_row_nbytes = out_w * stick_nbytes;
+    uint32_t reader_image_rows_per_core = (in_image_rows_per_core + is_reader) / 2;
+    uint32_t writer_image_rows_per_core = in_image_rows_per_core / 2;
+    uint32_t image_row_begin = is_reader ? 0 : reader_image_rows_per_core;
+    uint32_t image_row_end = is_reader ? reader_image_rows_per_core : in_image_rows_per_core;
+    uint32_t l1_read_addr = get_read_ptr(in_cb_id); //+ image_row_begin * in_image_row_nbytes;
+    constexpr uint32_t in_scalar_cb_id = tt::CB::c_in4;
+
+    // assuming shard begins with a new row. TODO: generalize?
+    float scale_h_inv = 1.0f / scale_h;
+    float scale_w_inv = 1.0f / scale_w;
+    for (uint32_t image_row = 0 ; image_row < in_image_rows_per_core * scale_h; ++image_row){
+        for(uint32_t j=0; j<  in_w * scale_w; j++){
+            cb_reserve_back(out_cb_id, 4);
+            cb_reserve_back(in_scalar_cb_id, 1);
+
+            uint32_t l1_write_addr = get_write_ptr(out_cb_id);
+
+            float y = (float)(image_row + 0.5f) * (float)scale_h_inv - 0.5f + read_offset; //need to change
+            float x = (float)(j + 0.5f) * (float)scale_w_inv -0.5f;
```

Comment:
This looks independent of the loop -- just calculate the `x` values once outside?